File: gtsam-rs/sys/src/nonlinear/levenberg_marquardt_optimizer.cc
#include "levenberg_marquardt_optimizer.h"
#include <algorithm>
#include <memory>

namespace gtsam {

std::unique_ptr<LevenbergMarquardtOptimizer>
new_levenberg_marquardt_optimizer(const NonlinearFactorGraph &graph,
                                  const Values &initial_values,
                                  const LevenbergMarquardtParams &params) {

    std::cout << "Factor Graph:\n"<< std::endl;
    graph.print();

    return std::make_unique<LevenbergMarquardtOptimizer>(graph, initial_values,
                                                        params);
}

std::unique_ptr<Values> optimize_safely(LevenbergMarquardtOptimizer &optimizer)
{
    Values estimate = optimizer.optimizeSafely();

    // std::cout << "LM VALUES ALL: ";
    // estimate.print();

    // std::cout << "GET FACTORS! ";
    // optimizer.getFactorsUnsafe().print();

    std::cout << "LM DETAILS: ";
    optimizer.print();


    return std::make_unique<Values>(estimate);
}

} // namespace gtsam

File: gtsam-rs/src/nonlinear/levenberg_marquardt_optimizer.rs
use cxx::UniquePtr;

use super::{
    levenberg_marquardt_params::LevenbergMarquardtParams,
    nonlinear_factor_graph::NonlinearFactorGraph,
    values::Values,
};

pub struct LevenbergMarquardtOptimizer {
    inner: UniquePtr<::sys::LevenbergMarquardtOptimizer>,
}

impl LevenbergMarquardtOptimizer {
    pub fn new(
        graph: &NonlinearFactorGraph,
        initial_values: &Values,
        params: &LevenbergMarquardtParams,
    ) -> Self {
        Self {
            inner: ::sys::new_levenberg_marquardt_optimizer(
                &graph.inner,
                &initial_values.inner,
                &params.inner,
            ),
        }
    }

    pub fn optimize_safely(&mut self) -> Values {
        Values {
            inner: ::sys::optimize_safely(self.inner.pin_mut()),
        }
    }
}
